org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/packageview/PackageExplorerContentProvider.java

```diff
@@ -127,6 +128,20 @@ protected Object getViewerInput() {
 
 	@Override
 	public void elementChanged(final ElementChangedEvent event) {
+		IJavaElementDelta delta= event.getDelta();
+		if (Display.findDisplay(Thread.currentThread()) == null) {
+			processDelta(delta);
+		} else {
+			Job job= Job.create(PackagesMessages.PackageExplorerContentProvider_update_job_description, m -> {
```

Comment:
As said, that shouldn't be just "fire and forget" with N jobs running in parallel. it should be one job that has a queue of deltas to process in exact order the deltas were fired. This code should put delta to the queue and single job instance should be triggered to process it.

Reply:
As this is the UI thread there are never n-jobs only one... Delta firing has no real order also right now because each thread uses its own DeltaProcessor....